File: router/company/company.py
from collections.abc import Callable
from typing import List, Optional

from fastapi import APIRouter, Query, Depends, Body
from sqlmodel import Session

from instance.dependency.dependency import get_company_service, api_security, \
    get_end_of_day_change_overview_orchestrator, get_company_overview_search_service, get_session, \
    get_company_details_orchestrator
from model.end_of_day_change_overview.end_of_day_change_overview import EndOfDayChangeOverview
from orchestrator.company_details_orchestrator.company_details_orchestrator import CompanyDetailsOrchestrator
from orchestrator.eodhd.end_of_day_change_overview_orchestrator import EndOfDayChangeOverviewOrchestrator
from schema.company.company import CompanyOverviewSchema, CompanyDetailsSchema
from schema.company.company_search.company_search_sql import Criterion
from service.company.company_overview_search_service import CompanyOverviewSearchService
from service.company.company_service import CompanyService

company_router = APIRouter(prefix="/company")


@company_router.get("/overview")
async def get_company_overview(company_ids: str = Query(...),
                               company_service: CompanyService = Depends(get_company_service),
                               session: Session = Depends(get_session),
                               security: Callable = Depends(api_security)) -> List[CompanyOverviewSchema]:
    return_ = company_service.company_overview(list(map(lambda id_: int(id_), company_ids.split(','))))
    session.close()

    return return_


@company_router.get("/details")
async def get_company_details(company_id: int = Query(...), session: Session = Depends(get_session),
                              company_details_orchestrator: CompanyDetailsOrchestrator = Depends(
                                  get_company_details_orchestrator),
                              security: Callable = Depends(api_security)) -> Optional[CompanyDetailsSchema]:
    return_ = await company_details_orchestrator.by_company_id(company_id)
    session.close()

    return return_


@company_router.get("/end-of-day-change-overview")
async def get_end_of_day_change_overview(ticker_id: int = Query(...),
                                         end_of_day_change_overview_orchestrator: EndOfDayChangeOverviewOrchestrator = Depends(
                                             get_end_of_day_change_overview_orchestrator),
                                         security: Callable = Depends(api_security),
                                         session: Session = Depends(get_session)) -> EndOfDayChangeOverview:
    return_ = await end_of_day_change_overview_orchestrator.by_ticker_id(ticker_id)
    session.close()

    return return_


@company_router.post("/search")
async def search(query: Optional[str] = Query(default=None), criteria: Optional[List[Criterion]] = Body(default=None),
                 company_overview_search_service: CompanyOverviewSearchService = Depends(
                     get_company_overview_search_service), session: Session = Depends(get_session),
                 security: Callable = Depends(api_security)):  #
    return_ = company_overview_search_service.search(query=query, criteria=criteria)
    session.close()

    return return_

File: nexus/src/actions/financial/company.ts
import {fastApiClient} from "@/instance/axios-instance";
import {company_details_type, company_overview_type} from "@/schema/schema";
import {criterion_type} from "@/schema/criterion-schema";

export async function search(query: string | undefined = undefined, criteria: criterion_type[] | undefined = undefined): Promise<company_overview_type[]>
{
    try
    {
        const response_ = await fastApiClient.post<company_overview_type[]>("/company/search", criteria, {
            params:
                {
                    query: query,
                },
        });

        return response_.data;
    }
    catch (err)
    {
        return [];
    }
}

export async function retrieveCompanyOverview(company_ids: number[]): Promise<company_overview_type[]>
{
    try
    {

        const response_ = await fastApiClient.get<company_overview_type[]>("/company/overview", {
            params:
                {
                    company_ids: company_ids.join(',')
                }
        });

        return response_.data;
    }
    catch (err)
    {
        console.log(err);

        return [];
    }
}

export async function retrieveCompanyOverviewSingle(company_id: number): Promise<company_overview_type | undefined>
{
    const array_ = await retrieveCompanyOverview([company_id]);

    if (array_.length)
    {
        return array_[0];
    }
    else
        return undefined;
}

export async function retrieveCompanyDetails(company_id: number): Promise<company_details_type | undefined>
{
    try
    {
        const response_ = await fastApiClient.get<company_details_type>("/company/details", {
            params:
                {
                    company_id,
                }
        });

        return response_.data;
    }
    catch (err)
    {
        console.log(err);

        return undefined;
    }

}
